cpp/ql/src/semmle/code/cpp/ir/dataflow/DefaultTaintTracking.qll

```diff
@@ -4,25 +4,17 @@ private import semmle.code.cpp.ir.dataflow.DataFlow
 private import semmle.code.cpp.ir.IR
 
 /**
- * A predictable expression is one where an external user can predict
+ * A predictable instruction is one where an external user can predict
  * the value. For example, a literal in the source code is considered
  * predictable.
  */
-// TODO: Change to use Instruction instead of Expr. Naive attempt breaks
-// TaintedAllocationSize qltest.
-private predicate predictable(Expr expr) {
-  expr instanceof Literal
+private predicate predictableInstruction(Instruction instr) {
+  instr instanceof ConstantInstruction
```

Comment:
Should this be `ConstantValueInstruction`?

Reply:
Good question. Neither class has documentation. Based on skimming through `TranslatedExpr.qll`, it looks to me like all instances of `StringConstantInstruction` are also instances of `ConstantValueInstruction` even though there is no subclass relationship between them. But what is the intended future evolution of `ConstantValueInstruction`? Instructions will be instances of this class if they _have_ a constant on them, not just if they _are_ constants. I'm not convinced that any future instruction that _has_ a constant will necessarily _be_ a constant. If that's the case, it should certainly be documented on `ConstantValueInstruction` and on `TranslatedElement.getInstructionConstantValue`.

Reply:
I'll merge as-is, then.

@dbartol what are your thoughts on `ConstantValueInstruction` vs `ConstantInstruction` and `StringConstantInstruction`?